Add SEP for ledger metadata storage (#1677)
stellar/stellar-protocol: ecosystem/sep-datastore.md

SEP: To Be Assigned
Title: Ledger Metadata Storage
Author: Tamir Sen <@tamirms>
Status: Draft
Created: 2025-03-11
Version: 0.1.0
```

## Simple Summary

A standard for how [`LedgerCloseMeta`](https://github.com/stellar/stellar-xdr/blob/v22.0/Stellar-ledger.x#L539-L545)
objects should be stored so that ledgers can be easily and efficiently ingested by downstream systems.

## Dependencies

None.

## Motivation

[galexie](https://github.com/stellar/go/tree/master/services/galexie) is a service which publishes
[`LedgerCloseMeta`](https://github.com/stellar/stellar-xdr/blob/v22.0/Stellar-ledger.x#L539-L545) XDR objects to a GCS
(Google Cloud Storage) bucket. However, the data format and layout of the XDR objects are not formally documented. This
SEP aims to provide a comprehensive specification for storing LedgerCloseMeta objects, enabling third-party developers
to build compatible data stores and clients for retrieving ledger metadata.

## Specification

The data store is a key-value store where:

- **Keys** are strings following a specific hierarchical format.
- **Values** are binary blobs representing compressed `LedgerCloseMetaBatch` XDR values.

The key-value store must support:

> - Efficient point lookups for arbitrary keys.

Reviewer: ```suggestion
- Efficient random access lookups on arbitrary keys.
```

- Listing keys in lexicographic order, optionally filtered by a prefix.

> Examples of compatible key-value stores include Google Cloud Storage (GCS) and Amazon S3.

Reviewer: maybe mention O/S File System here as well, to highlight it's not cloud dependent?

---

### Value Format

Each value in the key-value store is the compressed binary encoding of the following XDR structure:

```c++
// Batch of ledgers along with their transaction metadata
struct LedgerCloseMetaBatch
{
    // starting ledger sequence number in the batch
    uint32 startSequence;

    // ending ledger sequence number in the batch
    uint32 endSequence;

    // Ledger close meta for each ledger within the batch
    LedgerCloseMeta ledgerCloseMetas<>;
};
```

- A LedgerCloseMetaBatch represents a contiguous range of one or more consecutive ledgers.

> - All batches in a data store instance contain the same number of ledgers.

Reviewer: ```suggestion
- For data store consistency, all batches must contain the same quantity of ledgers, adhering to `ledgersPerBatch`.
```
I think can move this comment on constraint down below to line 113 to be closer to `ledgersPerBatch`

---

### Key Format

> Keys follow a hierarchical directory structure. The root directory is `/ledgers``, and subdirectories represent

Reviewer 2: I don't think we have a fixed prefix (`/ledgers`) for the root directory.

partitions. Each partition contains a fixed number of batches:

```

> /ledgers/<partition>/<batch>.xdr

Reviewer 2: ```suggestion
/ledgers/<partition>/<batch>.xdr.zstd
```

```

If the partition size is 1, the partition is omitted, resulting in:

```

> /ledgers/<batch>.xdr

Reviewer 2: ```suggestion
/ledgers/<batch>.xdr.zstd
```

```

#### Partition Format:

```go

> fmt.Sprintf("%08X--%d-%d/", math.MaxUint32-partitionStartLedgerSequence, partitionStartLedgerSequence, partitionEndLedgerSequence)

Reviewer: this is fine, may be a stretch for people to visualize the output, for consideration, a mask that demonstrates same format with literal placeholder name/type:
`<partition_start_32bit_hex>--<partition_start_int>-<partition_end_int>`

```

#### Batch Format:

```go
 fmt.Sprintf("%08X--%d-%d.xdr", math.MaxUint32-batchStartLedgerSequence, batchStartLedgerSequence, batchEndLedgerSequence)
```

If the batch size is 1, the format simplifies to:

```go
 fmt.Sprintf("%08X--%d.xdr", math.MaxUint32-batchStartLedgerSequence, batchStartLedgerSequence)
```

---

### Configuration File

The data store includes a configuration JSON object stored under the key `/config.json``. This file contains the
following properties:

- `networkPassphrase` - (string) the passphrase for the Stellar network associated with the ledgers.
- `compression` - (string) the compression algorithm used to compress ledger objects (currently only
  [`zstd`]([https://facebook.github.io/zstd/) is supported).
- `ledgersPerBatch` - (integer) the number of ledgers bundled into each `LedgerCloseMetaBatch`.
- `batchesPerPartition` - (integer) the number of batches in a partition.

#### Example Configuration:

```json
{
  "networkPassphrase": "Public Global Stellar Network ; September 2015",
  "compression": "zstd",
  "ledgersPerBatch": 2,
  "batchesPerPartition": 8
}
```

---

### Example Key Structure

> Below is an example list of keys for ledger batches based on the configuration above:

Reviewer: nice!

## Preamble

```
SEP: To Be Assigned
Title: Ledger Metadata Storage
Author: Tamir Sen <@tamirms>
Status: Draft
Created: 2025-03-11

> Version: 0.2.0

Reviewer 3: Could you add a Discussion parameter here, and link to the GitHub Discussion?
